### DISCUSSION TITLE
Diesel joins use "explicit join" syntax, degrading query planning

### DISCUSSION THREAD

**Discussion by Ten0** [diesel-rs/diesel#3216] [Ideas]
As it turns out, Diesel's way to actually write joins has an impact on the query planner (of Postgres at least).

A query like the following:
https://docs.diesel.rs/master/diesel/prelude/trait.QueryDsl.html#method.inner_join
```rust
users.inner_join(posts.inner_join(comments))
```

Does not precisely generate, as documented:
```sql
SELECT * FROM users
    INNER JOIN posts ON posts.user_id = users.id
    INNER JOIN comments ON comments.post_id = posts.id
```

Instead, it generates something that is morally equivalent:
```sql
SELECT * FROM users
	INNER JOIN (
		posts
		INNER JOIN comments ON comments.post_id = posts.id
	) ON posts.user_id = users.id
```

However, while these two queries produce the same results, they are not handled exactly the same way by e.g. postgres: the latter is considered an ["explicit join"](https://www.postgresql.org/docs/current/explicit-joins.html), and if the number of joins exceeds `8` (by default, `join_collapse_limit` parameter), postgres will not use the usual query planning to choose the best plan, as it would if the query was flat, and will instead use the joining order that is "explicitly specified".

This will result in queries such as (of course, bigger is required):
```sql
SELECT * FROM users
	INNER JOIN (
		posts
		INNER JOIN comments ON comments.post_id = posts.id
	) ON posts.user_id = users.id
	WHERE users.id = 73823
```
to go through all the posts and comments tables to build the entire
```sql
	(
		posts
		INNER JOIN comments ON comments.post_id = posts.id
	)
```
collection in memory/disk, and then hashjoin that to users later, instead of just using the relevant indexes.

I imagine that this is not intended behavior (in that the way we build queries is just a way that works well within the query builder but this is not an intended effect).

I think this should either get strongly documented, worked around or fixed somehow. :)

Possible paths:
1. Document this behavior and that people should probably set the `join_collapse_limit`/`from_collapse_limit` settings on their database (to higher values than `8`)
2. Do it ourselves by default as we initialize the connection
3. Rewrite the join part of the query building so that joins are written flat

Side note:
 - I have not checked how this writing impacts other backends than Postgres.

**Comment by weiznich** [comment]
Thanks for the detailed write up ❤️ 

I would likely go with the first "solution" for now. I wouldn't touch database defaults for query planing in the default setup as I assume those defaults are there for some reason.

**Reply by Ten0** [reply]
> I wouldn't touch database defaults for query planing in the default setup as I assume those defaults are there for some reason

I believe that reason is that they assumed that if people wrote using the explicit syntax, they actually *wanted* to specify the plan, but if query planning search is going to be easily exhaustive anyway they still flatten because they can find the best. (If not however, they'd rather use the user-specified plan instead of the genetic query planner's plan.)

The assumption that "if people wrote using the explicit syntax, they actually *wanted* to specify the plan" is incorrect in our case though.
